test/pylib/cpp/boost/prepare_instance.py

```diff
@@ -0,0 +1,116 @@
+            # so it will finish before the file is created
+            time.sleep(0.5)
+
+    def cleanup(self) -> None:
+        """
+        Fake method to have the same interfaces with Controller class.
+        """
+        pass
+
+
+class PrepareMainProcessEnv:
+    """
+    A class responsible for starting additional services needed by tests.
+
+    It starts up a Minio server and an S3 mock server.
+    The environment settings are saved to a file for later consumption by child processes.
+    """
+    def __init__(self, temp_dir: Path, modes: list[str], env_file: Path):
+        self.temp_dir = temp_dir
+        pytest_dirs = [self.temp_dir / mode / 'pytest' for mode in modes]
+        for directory in [self.temp_dir, *pytest_dirs]:
+            if not directory.exists():
+                directory.mkdir()
+        self.env_file = env_file
+        hosts = HostRegistry()
+        self.loop = asyncio.new_event_loop()
+        address = self.loop.run_until_complete(hosts.lease_host())
+        self.ms = MinioServer(self.temp_dir, '127.0.0.1', LogPrefixAdapter(logging.getLogger('minio'), {'prefix': 'minio'}))
+        self.mock_s3 = MockS3Server(address, 2012,
+                                    LogPrefixAdapter(logging.getLogger('s3_mock'), {'prefix': 's3_mock'}))
+
+    def prepare(self) -> None:
+        """
+        Start the S3 mock server and MinIO for the tests.
+        Create a file with environment variables for connecting to them.
+        """
+        tasks = [self.loop.create_task(self.ms.start()), self.loop.create_task(self.mock_s3.start())]
+        self.loop.run_until_complete(asyncio.gather(*tasks))
+        envs = self.ms.get_envs_settings()
+        envs.update(self.mock_s3.get_envs_settings())
+        with open(self.env_file, 'w') as file:
+            for key, value in envs.items():
+                file.write(f"{key}={value}\n")
+
+    def cleanup(self) -> None:
+        """
+        Stop the S3 mock server and MinIO
+        Remove the file with environment variables to not mess for consecutive runs.
+        """
+        tasks = [self.loop.create_task(self.ms.stop()), self.loop.create_task(self.mock_s3.stop())]
+        self.loop.run_until_complete(asyncio.gather(*tasks))
+        if os.path.exists(self.env_file):
+            self.env_file.unlink()
+
+def get_prepare_instance(temp_dir: Path, is_worker: bool, modes: list[str]) -> PrepareChildProcessEnv | PrepareMainProcessEnv:
+    """
+    xdist helps to execute test in parallel.
+    For that purpose it creates one main controller and workers.
+    Pytest itself doesn't know if it's a worker or controller, so it will execute all fixtures and methods.
+    Tests need S3 mock server and minio to start only once for the whole run, since they can share the one instance and
+    share the environment variables with workers.
+
+    So the part of starting the servers executes on non-workers' machines.
```

Comment:
I don't know what you mean "non-worker machines". Maybe you mean non-worker processes, or, basically - all xdist processes but the first process?

Reply:
Yes, this should be a process, not a machine. I'll update the comment.